File: very hard/Shadows of the Knight - Episode 2/Perl/Shadows of the Knight - Episode 2.pl
use strict;
use warnings;
use 5.32.1;
use List::Util qw(min max);

select(STDOUT); $| = 1; # DO NOT REMOVE

###################################################################################################################
# Function narrow_dimension:                                                                                      #
#   Updates the search interval in one dimension                                                                  #
#   based on the feedback and the comparison between                                                              #
#   the previous position (x0 or y0) and the current position (x or y).                                           #
#                                                                                                                 #
# Parameters:                                                                                                     #
#   $prev : previous position on the axis (x0 or y0)                                                              #
#   $current : current position on the axis (x or y)                                                              #
#   $min : current lower bound (x_min or y_min)                                                                   #
#   $max : current upper bound (x_max or y_max)                                                                   #
#   $info : feedback ("WARMER", "COLDER", "SAME", "UNKNOWN")                                                      #
#                                                                                                                 #
# For SAME, if (x0+x) (or y0+y) is even, the bomb is exactly in the middle.                                       #
# For WARMER/COLDER, we deduce which side of the interval to keep                                                 #
#   - WARMER: if moved closer to bomb, keep nearer half                                                            #
#   - COLDER: if moved further, discard nearer half                                                               #
###################################################################################################################
sub narrow_dimension {
    my ($prev, $current, $min, $max, $info) = @_;
    my $sum = $prev + $current;

    if ($info eq 'SAME') {
        if ($sum % 2 == 0) {
            my $middle = int($sum / 2);
            $min = $middle;
            $max = $middle;
        }
    }
    elsif ($info eq 'WARMER') {
        if ($current > $prev) {
            my $new_lower = int($sum / 2) + 1;
            $min = $new_lower if $new_lower > $min;
        } elsif ($current < $prev) {
            my $new_upper = ($sum % 2 == 0) ? int($sum / 2) - 1 : int($sum / 2);
            $max = $new_upper if $new_upper < $max;
        }
    }
    elsif ($info eq 'COLDER') {
        if ($current > $prev) {
            my $new_upper = ($sum % 2 == 0) ? int($sum / 2) - 1 : int($sum / 2);
            $max = $new_upper if $new_upper < $max;
        } elsif ($current < $prev) {
            my $new_lower = int($sum / 2) + 1;
            $min = $new_lower if $new_lower > $min;
        }
    }

    return ($min, $max);
}

##############################################################
# Function narrow:                                           #
# Applies narrow_dimension on x as long as                   #
# the horizontal interval is not reduced to a single value,  #
# then on y if x is already determined.                      #
##############################################################
sub narrow {
    my ($x0, $y0, $x, $y, $x_min, $x_max, $y_min, $y_max, $info) = @_;

    print STDERR "narrow: x0: $x0 y0: $y0 x_min: $x_min x_max: $x_max y_min: $y_min y_max: $y_max info: $info\n";

    if ($x_min != $x_max) {
        ($x_min, $x_max) = narrow_dimension($x0, $x, $x_min, $x_max, $info);
    } elsif ($y_min != $y_max) {
        ($y_min, $y_max) = narrow_dimension($y0, $y, $y_min, $y_max, $info);
    }

    print STDERR "narrow : x_min: $x_min x_max: $x_max y_min: $y_min y_max: $y_max\n";
    return ($x_min, $x_max, $y_min, $y_max);
}

# Read initial input
my $tokens;
chomp($tokens = <STDIN>);
my ($w, $h) = split(/ /, $tokens);
chomp(my $n = <STDIN>);
chomp($tokens = <STDIN>);
my ($x0, $y0) = split(/ /, $tokens);

my $x = $x0;
my $y = $y0;
my ($x_min, $x_max) = (0, $w - 1);
my ($y_min, $y_max) = (0, $h - 1);

# Main loop
while (1) {
    chomp(my $info = <STDIN>);

    ($x_min, $x_max, $y_min, $y_max) = narrow($x0, $y0, $x, $y, $x_min, $x_max, $y_min, $y_max, $info);

    $x0 = $x;
    $y0 = $y;

    if ($x_min != $x_max) {
        if ($x0 == 0 && $x_max + $x_min + 1 != $w) {
            $x = int((3 * $x_min + $x_max) / 2) - $x0;
        } elsif ($x0 == $w - 1 && $x_max + $x_min + 1 != $w) {
            $x = int(($x_min + 3 * $x_max) / 2) - $x0;
        } else {
            $x = $x_max + $x_min - $x0;
        }

        $x = $x0 + 1 if $x == $x0;
        $x = 0 if $x < 0;
        $x = $w - 1 if $x > $w - 1;
    } else {
        if ($x != $x_min) {
            $x = $x_min;
            $x0 = $x;
            print "$x $y\n";
            chomp($info = <STDIN>);
        }

        if ($y_min != $y_max) {
            if ($y0 == 0 && $y_max + $y_min + 1 != $h) {
                $y = int((3 * $y_min + $y_max) / 2) - $y0;
            } elsif ($y0 == $h - 1 && $y_max + $y_min + 1 != $h) {
                $y = int(($y_min + 3 * $y_max) / 2) - $y0;
            } else {
                $y = $y_max + $y_min - $y0;
            }

            $y = $y + 1 if $y == $y0;
            $y = 0 if $y < 0;
            $y = $h - 1 if $y > $h - 1;
        } else {
            $y = $y_min;
        }
    }

    print "$x $y\n";
}

File: very hard/Shadows of the Knight - Episode 2/OCaml/Shadows of the Knight - Episode 2.ml
(* ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////// *)
(* // Function narrow_dimension:                                                                                      *)
(* //   Updates the search interval in one dimension                                                                  *)
(* //   based on the feedback and the comparison between                                                              *)
(* //   the previous position (x0 or y0) and the current position (x or y).                                           *)
(* //                                                                                                                 *)
(* // Parameters:                                                                                                     *)
(* //   $1 : previous position on the axis (x0 or y0)                                                                 *)
(* //   $2 : current position on the axis (x or y)                                                                    *)
(* //   $3 : current lower bound (x_min or y_min)                                                                     *)
(* //   $4 : current upper bound (x_max or y_max)                                                                     *)
(* //   $5 : feedback ("WARMER", "COLDER", "SAME", "UNKNOWN")                                                         *)
(* //                                                                                                                 *)
(* // For SAME, if (x0+x) (or y0+y) is even, the bomb is                                                              *)
(* // exactly in the middle.                                                                                          *)
(* //                                                                                                                 *)
(* // For WARMER/COLDER, we deduce which side of the interval to keep:                                                *)
(* //   - If x > x0 and WARMER, the bomb is to the right of the middle,                                               *)
(* //     so new_min = ((x0+x)/2) + 1.                                                                                *)
(* //   - If x > x0 and COLDER, the bomb is to the left, so new_max = (x0+x)/2 - (if even) or = (x0+x)/2 (if odd).    *)
(* //                                                                                                                 *)
(* // The same reasoning applies for y.                                                                               *)
(* ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////// *)
let narrow_dimension prev current min_ max_ info =
  let sum = prev + current in
  match info with
  | "SAME" ->
      if sum mod 2 = 0 then
        let middle = sum / 2 in
        (middle, middle)
      else (min_, max_)
  | "WARMER" ->
      if current > prev then
        let new_lower = (sum / 2) + 1 in
        (max min_ new_lower, max_)
      else if current < prev then
        let new_upper = if sum mod 2 = 0 then (sum / 2) - 1 else sum / 2 in
        (min_, min max_ new_upper)
      else (min_, max_)
  | "COLDER" ->
      if current > prev then
        let new_upper = if sum mod 2 = 0 then (sum / 2) - 1 else sum / 2 in
        (min_, min max_ new_upper)
      else if current < prev then
        let new_lower = (sum / 2) + 1 in
        (max min_ new_lower, max_)
      else (min_, max_)
  | _ -> (min_, max_)

(* //////////////////////////////////////////////////////////////// *)
(* // Function narrow:                                           // *)
(* // Applies narrow_dimension on x as long as                   // *)
(* // the horizontal interval is not reduced to a single value,  // *)
(* // then on y if x is already determined.                      // *)
(* //////////////////////////////////////////////////////////////// *)
let narrow x0 y0 x y x_min x_max y_min y_max info =
  prerr_endline (Printf.sprintf "narrow: x0: %d y0: %d x_min: %d x_max: %d y_min: %d y_max: %d info: %s" x0 y0 x_min x_max y_min y_max info);
  let x_min, x_max, y_min, y_max =
    if x_min <> x_max then
      let x_min', x_max' = narrow_dimension x0 x x_min x_max info in
      (x_min', x_max', y_min, y_max)
    else if y_min <> y_max then
      let y_min', y_max' = narrow_dimension y0 y y_min y_max info in
      (x_min, x_max, y_min', y_max')
    else (x_min, x_max, y_min, y_max)
  in
  prerr_endline (Printf.sprintf "narrow : x_min: %d x_max: %d y_min: %d y_max: %d" x_min x_max y_min y_max);
  (x_min, x_max, y_min, y_max)

let () =
  let w, h = Scanf.sscanf (input_line stdin) " %d %d" (fun w h -> w, h) in
  let _ = int_of_string (input_line stdin) in
  let x0_init, y0_init = Scanf.sscanf (input_line stdin) " %d %d" (fun x y -> x, y) in

  let rec loop x0 y0 x y x_min x_max y_min y_max =
    let info = input_line stdin in
    let x_min, x_max, y_min, y_max = narrow x0 y0 x y x_min x_max y_min y_max info in

    let x0, y0 = x, y in
    let x =
      if x_min <> x_max then
        let x =
          if x0 = 0 && x_max + x_min + 1 <> w then
            (3 * x_min + x_max) / 2 - x0
          else if x0 = w - 1 && x_max + x_min + 1 <> w then
            (x_min + 3 * x_max) / 2 - x0
          else
            x_max + x_min - x0
        in
        let x = if x = x0 then x0 + 1 else x in
        let x = max 0 (min (w - 1) x) in
        x
      else
        if x <> x_min then begin
          let x = x_min in
          let x0 = x in
          Printf.printf "%d %d\n%!" x y;
          let _ = input_line stdin in
          x
        end else x
    in

    let y =
      if x_min = x_max then
        if y_min <> y_max then
          let y =
            if y0 = 0 && y_max + y_min + 1 <> h then
              (3 * y_min + y_max) / 2 - y0
            else if y0 = h - 1 && y_max + y_min + 1 <> h then
              (y_min + 3 * y_max) / 2 - y0
            else
              y_max + y_min - y0
          in
          let y = if y = y0 then y + 1 else y in
          let y = max 0 (min (h - 1) y) in
          y
        else y_min
      else y
    in

    Printf.printf "%d %d\n%!" x y;
    loop x0 y0 x y x_min x_max y_min y_max
  in

  let x = x0_init in
  let y = y0_init in
  let x_min = 0 and x_max = w - 1 in
  let y_min = 0 and y_max = h - 1 in
  loop x0_init y0_init x y x_min x_max y_min y_max
